### PR TITLE
volumebinding: scheduler queueing hints - StorageClass

### PR DESCRIPTION
**Pull request by bells17**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

kube-scheduler implements scheduling hints for the VolumeBinding plugin.
The scheduling hints allow the scheduler to determine whether to retry or skip scheduling a Pod based on the changes made to the StorageClass resource referenced by the plugin.

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/kubernetes/issues/118893
KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/4247-queueinghint/README.md
Base PR: https://github.com/kubernetes/kubernetes/pull/124939

#### Special notes for your reviewer:

<details><summary>Fields Impacting QueueingHintFn</summary>
<p>
<strong>PersistentVolume (PV)</strong> is not included in this table because it can undergo extensive changes when a conversion is performed by csi-translation-lib.
</p>

| resource | field | Referenced in PreFilter+Filter? | Admission Overwrite Prevention Config | Need to Check Changes in QHint? |
| ---- | ---- | ---- | --- | --- |
| StorageClass | .metadata.labels  | x | x | x |
| StorageClass | .metadata.annotations  | x | x | x |
| StorageClass | .provisioner  | o | o | x |
| StorageClass | .parameters  | x | o | x |
| StorageClass | .reclaimPolicy  | x | o | x |
| StorageClass | .mountOptions  | x | o | x |
| StorageClass | .allowVolumeExpansion  | x | x | x |
| StorageClass | .volumeBindingMode  | o | x | o |
| StorageClass | .allowedTopologies  | o | x | o |

ref(ja): https://zenn.dev/bells17/scraps/65bd6891012bdc

</details> 

#### Does this PR introduce a user-facing change?

```release-note
kube-scheduler implements scheduling hints for the VolumeBinding plugin.
The scheduling hints allow the scheduler to retry scheduling a Pod that was previously rejected by the VolumeBinding plugin only if a new resource referenced by the plugin was created or an existing resource referenced by the plugin was updated.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding.go [hunk 1]
```diff
@@ -123,32 +132,71 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
 	return events
 }
 
+func (pl *VolumeBinding) isSchedulableAfterStorageClassChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	oldSC, newSC, err := util.As[*storagev1.StorageClass](oldObj, newObj)
+	if err != nil {
+		return framework.Queue, err
+	}
+
```
Comment (carlory) [inline pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:140]: It's forbidden to update the volume binding mode of a storage class. If a storage class's binding mode isn't VolumeBindingWaitForFirstConsumer, when it is created or updated, the action itself doesn't make the pod become schedulable directly. If then some PVCs is bound by PVs due to the storage class (re-)created, it should be reflected in the PVC update event. so in this function, we should only take care of special storage classes with the VolumeBindingWaitForFirstConsumer binding mode.

Comment (carlory) [reply pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:140]: If we only consider VolumeBindingWaitForFirstConsumer, this PR will depend on the other hint func which handle PVC events. 

So let's merge another firstly once that PR is ready to be merged. @sanposhiho

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding.go [hunk 2]
```diff
@@ -123,32 +132,71 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
...
+		"StorageClass", klog.KObj(newSC),
+	)
+
+	pinfoList := getPVCInfoListFromPodVolumes(pod)
+	for _, pinfo := range pinfoList {
+		pvc, err := pl.PVCLister.PersistentVolumeClaims(pod.Namespace).Get(pinfo.pvcName)
+		if err != nil {
+			if pinfo.isEphemeral && apierrors.IsNotFound(err) {
+				err = fmt.Errorf("waiting for ephemeral volume controller to create the persistentvolumeclaim %q", pinfo.pvcName)
+				return framework.Queue, err
```
Comment (carlory) [inline pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:153]: Why return an error? If so, the pod will put into BackOffQ. It doesn't make pod become schedulable.

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding.go (hunk 1 above)
Comment (bells17) [reply pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:140]: Thank you for your suggestion. In that case, would you mind reviewing the following PR first, which adds QHint for PVC?
https://github.com/kubernetes/kubernetes/pull/124959

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding.go (hunk 2 above)
Comment (bells17) [reply pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:149]: I've tried committing the code content mentioned in the above comment.

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding.go [hunk 3]
```diff
@@ -123,6 +127,60 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
 	return events
 }
 
+func (pl *VolumeBinding) isSchedulableAfterStorageClassChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:130]: similar to https://github.com/kubernetes/kubernetes/pull/124961#discussion_r1615039425; if we can consider StorageClass update is not that frequent, it's not worthy for a fine but time-consuming filtering (using `PVCLister.PersistentVolumeClaims().Get()`).

Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:130]: As a first step, we should make it minimum, at least until we establish some observability around QHint and could ensure this level of fine filtering doesn't impact a large cluster negatively (https://github.com/kubernetes/kubernetes/issues/124566).

**Comment by bells17** [comment]
@sanposhiho Thank you for your review. I have made the necessary changes, so please take another look.

**Comment by bells17** [comment]
/close
https://github.com/kubernetes/kubernetes/pull/124961#discussion_r1620473466

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding.go [hunk 4]
```diff
@@ -123,6 +126,32 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
...
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:129]: mention that volume binding mode is immutable and we don't have to take the update there into consideration.

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding.go [hunk 5]
```diff
@@ -123,6 +126,32 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
...
+	}
+
+	logger = klog.LoggerWithValues(
+		logger,
+		"Pod", klog.KObj(pod),
+		"StorageClass", klog.KObj(newSC),
+	)
+
+	if oldSC == nil {
+		logger.V(4).Info("StorageClass was created")
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:142]:
```suggestion
		// No further filtering can be made for a creation event, 
		// and we just always return Queue.
		logger.V(5).Info("A new StorageClass was created, which could make a Pod schedulable")
```

**Review by sanposhiho — COMMENTED** [review]
Only small nits from me.
@carlory Can you also take a look?

**Comment by carlory** [comment]
/lgtm
for storage part

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding.go [hunk 6]
```diff
@@ -151,6 +152,34 @@ func (pl *VolumeBinding) isSchedulableAfterCSINodeChange(logger klog.Logger, pod
...
+
+	if oldSC == nil {
+		// No further filtering can be made for a creation event,
+		// and we just always return Queue.
+		logger.V(5).Info("A new StorageClass was created, which could make a Pod schedulable")
+		return framework.Queue, nil
+	}
+
+	if !apiequality.Semantic.DeepEqual(newSC.AllowedTopologies, oldSC.AllowedTopologies) {
+		logger.V(4).Info("StorageClass was created or updated, and changed AllowedTopologies", "AllowedTopologies", newSC.AllowedTopologies)
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:175]:
```suggestion
		logger.V(5).Info("StorageClass got an update in AllowedTopologies", "AllowedTopologies", newSC.AllowedTopologies)
```

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding.go [hunk 7]
```diff
@@ -151,6 +152,34 @@ func (pl *VolumeBinding) isSchedulableAfterCSINodeChange(logger klog.Logger, pod
...
+		return framework.Queue, nil
+	}
+
+	logger.V(5).Info("StorageClass was created or updated, but it doesn't make this pod schedulable")
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:179]:
```suggestion
	logger.V(5).Info("StorageClass was updated, but it doesn't make this pod schedulable")
```

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding.go (hunk 4 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:155]: You overlooked this ^

```suggestion
// isSchedulableAfterStorageClassChange checks whether an StorageClass event might make a Pod schedulable or not. 
// Any StorageClass addition and a StorageClass update to allowedTopologies 
// might make a Pod schedulable. 
// Note that an update to volume binding mode is not allowed and we don't have to consider while examining the update event.
func (pl *VolumeBinding) isSchedulableAfterStorageClassChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
```

**Review by sanposhiho — APPROVED** [review]
/lgtm
/approve

**Pull request merged**